### PR TITLE
Maven repository deploy to Sonatype OSSRH

### PR DESCRIPTION
**Pull request by Waldz**

### CODE REVIEW

File: bin/release_android_sonatype [hunk 1]
```diff
@@ -0,0 +1,27 @@
+#!/bin/bash
+
+###########################################################################
+# Uploads all created Android artifacts to the Bintray and JCenter
```
Comment (soffokl) [inline bin/release_android_sonatype]:
```suggestion
# Uploads all created Android artifacts to the Sonatype
```

File: bin/release_android_sonatype [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+# - https://bintray.com/mysterium/Node
```
Comment (soffokl) [inline bin/release_android_sonatype]: replace bintray to sonatype

File: bin/release_android_sonatype [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+#
+# Usage:
+#> bin/release_android <VERSION> <BINTRAY_USER_NAME> <BINTRAY_API_TOKEN>
```
Comment (soffokl) [inline bin/release_android_sonatype]: wrong script name.

File: bin/release_android_sonatype [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+source bin/helpers/output.sh
+
+VERSION=$1
+if [ -z "$VERSION" ]; then
+    print_error "Missing version"
+    exit 1
+fi
+sed -i -E "s|<version>.+</version>|<version>$VERSION</version>|g" bin/package/android/mvn.pom
+
```
Comment (soffokl) [inline bin/release_android_sonatype]: gpg key import and mvn binary check are missing.

File: bin/package/android/mvn.pom [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+  <organization>
+    <name>Mysterium Network</name>
+    <url>https://mysterium.network</url>
+  </organization>
+
+  <developers>
+    <developer>
+      <id>waldz</id>
+      <name>Valdas Petrulis</name>
```
Comment (tadovas) [inline bin/package/android/mvn.pom]: Shouldn't be Mysterium team as developer enough? Do we need to expose individuals too?

Comment (soffokl) [reply bin/package/android/mvn.pom]: Left only Mysterium team in a list of developers.

File: .travis.yml [hunk 6]
```diff
@@ -172,8 +172,11 @@ jobs:
       name: "Pushing release to hub.docker.com"
     - script:
       - bin/s3 sync s3://travis-$TRAVIS_BUILD_NUMBER/build-artifacts build/package
-      - bin/release_android "$TRAVIS_TAG" "$BINTRAY_USER" "$BINTRAY_TOKEN"
-      name: "Publish Android SDK to Bintray and JCenter"
+      - openssl aes-256-cbc -K $encrypted_2698a83519b3_key -iv $encrypted_2698a83519b3_iv -in bin/package/android/key.gpg.enc -out bin/package/android/key.gpg -d
```
Comment (Waldz) [inline .travis.yml]: Security issue here. We should inject this key from environment variables

File: bin/release_android [hunk 7]
```diff
@@ -17,22 +17,41 @@ if [ -z "$VERSION" ]; then
...
-fi
+URL="https://oss.sonatype.org/service/local/staging/deploy/maven2"
 
-TOKEN=$3
-if [ -z "$TOKEN" ]; then
-    print_error "Missing Bintray API token"
-    exit 1
+if [ "$VERSION" == "0.0.0-dev" ]; then
+    VERSION="0.0.0-SNAPSHOT"
+    URL=https://oss.sonatype.org/content/repositories/snapshots
```
Comment (Waldz) [inline bin/release_android]: URL should be injected from Travis, instead of of trying to guess it from version

File: bin/package/android/mvn.pom [hunk 8]
```diff
@@ -0,0 +1,53 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd" xmlns="http://maven.apache.org/POM/4.0.0"
+    xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance">
+  <modelVersion>4.0.0</modelVersion>
+
+  <groupId>network.mysterium</groupId>
+  <artifactId>mobile-node</artifactId>
+  <version>0.0.0-SNAPSHOT</version>
```
Comment (Waldz) [inline bin/package/android/mvn.pom]: `${VERSION}`

File: bin/release_android [hunk 9]
```diff
@@ -17,22 +17,41 @@ if [ -z "$VERSION" ]; then
...
 
-sed -i -E "s|<version>.+</version>|<version>$VERSION</version>|g" bin/package/android/pom.xml
+sed -i -E "s|<version>.+</version>|<version>$VERSION</version>|g" bin/package/android/mvn.pom
 
-URL=https://api.bintray.com/content/mysterium/Node/mobile-node/$VERSION/mysterium/network/mobile-node/$VERSION
+mvn gpg:sign-and-deploy-file -e -X \
+    -settings=bin/package/android/mvn.settings \
+    -Durl=$URL \
+    -DrepositoryId=ossrh \
+    -Dgpg.keyname=336826E2222A0B29 \
```
Comment (Waldz) [inline bin/release_android]: Does it need to be hardcoded?

File: bin/release_android [hunk 10]
```diff
@@ -17,22 +17,41 @@ if [ -z "$VERSION" ]; then
...
+    -DpomFile=bin/package/android/mvn.pom \
+    -Dfile=build/package/Mysterium.aar
+
+if [ "$VERSION" == "0.0.0-SNAPSHOT" ]; then
+    exit 0
```
Comment (Waldz) [inline bin/release_android]: Lets validate VERSION there it's parse from injected arguments

File: bin/release_android [hunk 11]
```diff
@@ -17,22 +17,41 @@ if [ -z "$VERSION" ]; then
...
-curl -T bin/package/android/pom.xml         -u$USER:$TOKEN "$URL/mobile-node-$VERSION.pom?override=1&publish=1"
-curl -T build/package/Mysterium.aar         -u$USER:$TOKEN "$URL/mobile-node-$VERSION.aar?override=1&publish=1"
-curl -T build/package/Mysterium-sources.jar -u$USER:$TOKEN "$URL/mobile-node-$VERSION-sources.jar?override=1&publish=1"
+RepoID=$(mvn nexus-staging:rc-list \
+    -settings=bin/package/android/mvn.settings \
+    -DnexusUrl=https://oss.sonatype.org/ \
+    -DserverId=ossrh \
+    | awk '/networkmysterium-/ {id=$2} END{print id}')
+
+mvn nexus-staging:rc-close \
```
Comment (Waldz) [inline bin/release_android]: You should install `mvn` into tooling container `builder`

File: .travis.yml [hunk 12]
```diff
@@ -172,8 +172,11 @@ jobs:
...
+      - gpg --import bin/package/android/key.gpg
+      - gpg --list-keys
+      - bin/release_android "$TRAVIS_TAG"
```
Comment (Waldz) [inline .travis.yml]: - Script could ask to inject variable for GPG signing key 
- Script names does not match with `release_android_sonatype`, please align

File: bin/release_android [hunk 13]
```diff
@@ -17,22 +17,41 @@ if [ -z "$VERSION" ]; then
     exit 1
 fi
 
-USER=$2
-if [ -z "$USER" ]; then
-    print_error "Missing Bintray API user"
-    exit 1
-fi
+URL="https://oss.sonatype.org/service/local/staging/deploy/maven2"
```
Comment (Waldz) [inline bin/release_android]: Looks like `SONATYPE_USERNAME` and `SONATYPE_PASSWORD` variables are missing

File: .travis.yml (hunk 6 above)
Comment (soffokl) [reply .travis.yml]: This is the Travis way to provide secure keys.
https://docs.travis-ci.com/user/encrypting-files/
What is the security issue?

File: .travis.yml (hunk 12 above)
Comment (soffokl) [reply .travis.yml]: The script is called `bin/release_android`.

File: bin/release_android (hunk 13 above)
Comment (soffokl) [reply bin/release_android]: Added it.

File: bin/release_android (hunk 10 above)
Comment (zolia) [reply bin/release_android]: Do we really? Its CI script not exposed to any running service.

File: bin/release_android (hunk 11 above)
Comment (zolia) [reply bin/release_android]: Will it be used anywhere else? Why to install it?

File: bin/release_android (hunk 7 above)
Comment (soffokl) [reply bin/release_android]: In this case, we will guess it in Travis. I just wanted to keep related things together.

File: bin/release_android (hunk 10 above)
Comment (soffokl) [reply bin/release_android]: I didn't get it, what should be validated?

File: bin/release_android (hunk 9 above)
Comment (soffokl) [reply bin/release_android]: No, but we are committing this key, so no other options available.

File: bin/release_android (hunk 11 above)
Comment (soffokl) [reply bin/release_android]: How it's related to builder container? It's already too big.

File: bin/package/android/mvn.pom (hunk 8 above)
Comment (soffokl) [reply bin/package/android/mvn.pom]: Changed it.

File: bin/release_android (hunk 11 above)
Comment (Waldz) [reply bin/release_android]: So it's already on Travis Machine.. I thought it's not

File: bin/release_android (hunk 10 above)
Comment (Waldz) [reply bin/release_android]: I meant weird exit of script in the middle of script + no message. Would be nice improve

**Comment by tadovas** [comment]
A ticket with number 666, pushing android to sonatype OSSRH. World wants to tell us something :|

**Pull request merged**